libs/partners/google-vertexai/langchain_google_vertexai/functions_utils.py

```diff
@@ -0,0 +1,116 @@
+import asyncio
+from typing import List, Union
+
+from langchain.agents.agent import AgentOutputParser
+from langchain_community.utils.openai_functions import (
+    FunctionDescription,
+)
+from langchain_core.agents import AgentAction, AgentActionMessageLog, AgentFinish
+from langchain_core.messages import (
+    AIMessage,
+    BaseMessage,
+)
+from langchain_core.outputs import ChatGeneration, Generation
+from langchain_core.tools import Tool
+from langchain_core.utils.json_schema import dereference_refs
+from vertexai.preview.generative_models import (
+    FunctionDeclaration,
+)
+from vertexai.preview.generative_models import (
+    Tool as VertexTool,  # type: ignore
+)
+
+
+def format_tool_to_vertex_function(tool: Tool) -> FunctionDescription:
+    "Format tool into the Vertex function API."
+    if tool.args_schema:
+        schema = dereference_refs(tool.args_schema.schema())
+        schema.pop("definitions", None)
+        return {
+            "name": tool.name or schema["title"],
+            "description": tool.description or schema["description"],
+            "parameters": schema,
+        }
+    else:
+        return {
+            "name": tool.name,
+            "description": tool.description,
+            "parameters": {
+                "properties": {
+                    "__arg1": {"type": "string"},
+                },
+                "required": ["__arg1"],
+                "type": "object",
+            },
+        }
+
+
+def format_tools_to_vertex_tool(tools: List[Tool]) -> List[VertexTool]:
+    "Format tool into the Vertex Tool instance."
+    function_declarations = []
+    for tool in tools:
+        function_declarations.append(
+            FunctionDeclaration(**format_tool_to_vertex_function(tool))
+        )
+
+    return [VertexTool(function_declarations=function_declarations)]
+
+
+class VertexAIFunctionsAgentOutputParser(AgentOutputParser):
```

Comment:
@lkuligin Would you consider using a parser here from langchain if it's using the same specification as open-ai functions calling but has a more generic name? Or do you want a separate one to avoid depending on code?

Reply:
Agreed - would be great to recommend a standard implementation here as more providers offer function calling!

Reply:
let's merge this one then, wdyt:
https://github.com/langchain-ai/langchain/pull/15841